### PR TITLE
switch-op design

### PR DESCRIPTION
**Pull request by jacquesqiao**
fix: https://github.com/PaddlePaddle/Paddle/issues/8027

### CODE REVIEW

File: doc/design/scalar_switch_case_op_design.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+```python
+def piecewise_decay():
+  global_step = pd.Var(shape=[1], value=...)
+  learning_rate = pd.Var(shape=[1], value=...)
+
+  candition1 = pd.less_equal(global_step, 100000)
+  candition2 = pd.and(pd.less_then(100000, global_step), pd.less)
+
+  switch = ScalarSwitchOp()
+  with switch.case(candition1):
```
Comment (Superjomn) [inline doc/design/scalar_switch_case_op_design.md]: do these three `case` similar to `if-else` or just `if`?  a little vague.

if it is like

```python
if cond1: pass
elif cond2: pass
else: pass #cond3
```
the three `switch.case` seems no logic relation with each other.

if it is just three `if` like

```python
if cond1: pass
if cond2: pass
if cond3: pass
```
then they shouldn't share the same `Switch` instance.

Need some improvement to specify what logic it is in first glimpse.

Comment (jacquesqiao) [reply doc/design/scalar_switch_case_op_design.md]: yes, this is a problem, we have several condition and the relation is not clear, I will think about this.

Comment (jacquesqiao) [reply doc/design/scalar_switch_case_op_design.md]: after some search, I think this situation can be more accurately described by `cond` op in Lisp <https://www.cs.cmu.edu/Groups/AI/html/cltl/clm/node84.html>

image

and the interface can be:
```python
def piecewise_decay():
  global_step = pd.Var(shape=[1], value=...)
  learning_rate = pd.Var(shape=[1], value=...)

  candition1 = pd.less_equal(global_step, 100000)
  candition2 = pd.and(pd.less_then(100000, global_step), pd.less)

  cand_op = ScalarCandOp()
  with cand_op.case(candition1):
    pd.assign(learning_rate, 1.0)
  with cand_op.case(candition2):
    pd.assign(learning_rate, 0.5)
  with cand_op.default_case():
    pd.assign(learning_rate, 0.1)

  return learning_rate
```

there are some features for `cond` op:
- 1. the order of condition is useful, the op will find a suitable branch to run from up to down in order.
- 2. only one branch will run.

**Comment by wangkuiyi** [comment]
I filed [a PR](https://github.com/jacquesqiao/Paddle/pull/3) to rewrite this document, to correct English grammar, to make it concise and shorter.  Please review it @jacquesqiao . Thanks.

**Comment by jacquesqiao** [comment]
@wangkuiyi  thank you for your review and update!

File: doc/design/switch.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+b = fluid.Var(0)
+
+cond = fluid.switch()
+with cond.case(fluid.less_equal(a, 10)):
+    fluid.print("Case 1")
+with cond.case(fluid.larger(a, 0)):
+    fluid.print("Case 2")
+with cond.default():
+    fluid.print("Case 3")
+```
```
Comment (Superjomn) [inline doc/design/switch.md]: We need to scope `cond`s to underline that these three conds are in the same scope.

Currently, the cases should be adjacent, but the `with` syntax does not have that constraint semantically, so user tends to feel weird and write some wrong code.

what if user write some code like

```python
cond = fluid.switch()
with cond.case(case1):
    pass
with cond.case(case2):
    pass
b = a + c
with cond.case(cond3):
    pass
```
the `b = a + c` in the cases will break the global scope/continuous relations of a `switch`-cases.

Discussed with @jacquesqiao 

In lisp, a `cond` op is like
```lisp
;; cond has a scope to wrap all the following cases
(cond ((> a 2) "gt")
      ((= a 2) "eq")
      (t "lt")
) ;; cond
```
it seems better to write code as follows
```python
cond = fluid.switch()
with fluid.switch() as switch:
    '''
    A limitation inside a `switch`-block that no operators can insert between `case`s.  
    This rule is easier to check.
    '''
    with switch.case(fluid.less_equal(a, 10)):
           pass
    with switch.case(fluid.larger(a, 1)):
           pass
    switch()
```

@wangkuiyi @jacquesqiao

Comment (jacquesqiao) [reply doc/design/switch.md]: maybe need a cond.scope() as follow, one cond must and only has one scope.
```python
cond = fluid.switch()
with cond.scope():
    '''
    A limitation inside a `switch`-block that no operators can insert between `case`s.  
    This rule is easier to check.
    '''
    with switch.case(fluid.less_equal(a, 10)):
           pass
    with switch.case(fluid.larger(a, 1)):
           pass
```

**Pull request merged**